### PR TITLE
8332360: JVM hangs at exit when running on a uniprocessor

### PR DESCRIPTION
**Pull request by coleenp**
Restored the TryLock in the ReenterI (wait) case with a comment.

Tested with tier1-4.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8332360](https://bugs.openjdk.org/browse/JDK-8332360): JVM hangs at exit when running on a uniprocessor (**Bug** - P3)

### Reviewers
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**) ⚠️ Review applies to [5c1cf6b2](https://git.openjdk.org/jdk/pull/19293/files/5c1cf6b27e9fe83d5b05d9f11f68fed23bf9c1e5)
 * [Daniel D. Daugherty](https://openjdk.org/census#dcubed) (@dcubed-ojdk - **Reviewer**) ⚠️ Review applies to [5c1cf6b2](https://git.openjdk.org/jdk/pull/19293/files/5c1cf6b27e9fe83d5b05d9f11f68fed23bf9c1e5)
 * [Aleksey Shipilev](https://openjdk.org/census#shade) (@shipilev - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/19293/head:pull/19293` \
`$ git checkout pull/19293`

Update a local copy of the PR: \
`$ git checkout pull/19293` \
`$ git pull https://git.openjdk.org/jdk.git pull/19293/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 19293`

View PR using the GUI difftool: \
`$ git pr show -t 19293`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/19293.diff">https://git.openjdk.org/jdk/pull/19293.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/19293#issuecomment-2119384491)

### CODE REVIEW

File: src/hotspot/share/runtime/objectMonitor.cpp [hunk 1]
```diff
@@ -987,6 +987,13 @@ void ObjectMonitor::ReenterI(JavaThread* current, ObjectWaiter* currentNode) {
     guarantee(v == ObjectWaiter::TS_ENTER || v == ObjectWaiter::TS_CXQ, "invariant");
     assert(owner_raw() != current, "invariant");
 
+    // This thread has been notified to reacquire the lock.
```
Comment (dholmes-ora) [inline src/hotspot/share/runtime/objectMonitor.cpp:990]: Suggestion: s/to/so/

**Review by dholmes-ora — COMMENTED** [review]
Another possibility here could be to hoist the initial `TryLock` out of the loop. With the old, and now proposed again, code, the loop performs two successive tryLocks:
```
TryLock
TrySpin
park
TryLock
// loops
TryLock
TrySpin
park
TryLock
// loops
```
Of course the expectation is that if we are unparked then the next `TryLock` will succeed so looping like this would be somewhat pathological.

**Review by dholmes-ora — COMMENTED** [review]
Okay looks good. Restoring old code avoids any further risks.

Thanks

File: src/hotspot/share/runtime/objectMonitor.cpp [hunk 2]
```diff
@@ -987,6 +987,13 @@ void ObjectMonitor::ReenterI(JavaThread* current, ObjectWaiter* currentNode) {
     guarantee(v == ObjectWaiter::TS_ENTER || v == ObjectWaiter::TS_CXQ, "invariant");
     assert(owner_raw() != current, "invariant");
 
+    // This thread has been notified so reacquire the lock.
```
Comment (dcubed-ojdk) [inline src/hotspot/share/runtime/objectMonitor.cpp:990]: nit typo: s/so reacquire/so try to reacquire/

**Review by dcubed-ojdk — APPROVED** [review]
Thumbs up. One minor nit comment.

File: src/hotspot/share/runtime/objectMonitor.cpp (hunk 2 above)
Comment (coleenp) [reply src/hotspot/share/runtime/objectMonitor.cpp]: Fixed.

**Review by shipilev — APPROVED** [review]
Nice landmine! Fix looks good.

**Comment by tstuefe** [comment]
Nice analysis, @dholmes-ora . Should we do something about the inability to mimic a uniprocessor system? Possibly make is_MP dependent on sched_getaffinity?

**Comment by coleenp** [comment]
Thanks David, Dan and Aleksey.
/integrate

**Pull request closed without merging**